crates/win-api-wrappers/src/user.rs

```diff
@@ -0,0 +1,30 @@
+use windows::core::PWSTR;
+use windows::Win32::{Foundation::HINSTANCE, UI::WindowsAndMessaging::LoadStringW};
+
+pub fn load_string(hinstance: HINSTANCE, resource_id: u32) -> anyhow::Result<Option<String>, windows::core::Error> {
+    let mut resource_ptr: PWSTR = PWSTR::null();
+
+    // SAFETY:
+    // No preconditions. An invalid `hinstance` or `resource_id` will cause `LoadStringW` to return 0.
+    // We pass a null-initialized mutable reference to `resource_ptr`. This is safe because:
+    // - `resource_ptr` is a properly aligned, initialized `PWSTR` variable that is of sufficient length to hold a pointer.
+    // - if `cchBufferMax` is equal to 0, `LoadStringW` guarantees to write a valid pointer to the start of a resource string or leave it as null if no resource is found.
+    // The use of `std::mem::transmute` here provides compatibility between `&mut PWSTR` and the raw pointer type expected by `LoadStringW`.
+    // - This is sound because `&mut PWSTR` has the same memory layout as the pointer type expected by `LoadStringW`.
+    let length = unsafe { LoadStringW(hinstance, resource_id, std::mem::transmute(&mut resource_ptr), 0) };
```

Comment:
note: It’s suspicious to use transmute here, but I see that the signature in the [windows crate](https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/UI/WindowsAndMessaging/fn.LoadStringW.html) and the [Win32 doc](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-loadstringw) is not matching:

![image](https://github.com/user-attachments/assets/3cb8e20e-4224-4a7d-9880-ea956bf79738)
![image](https://github.com/user-attachments/assets/5526dd11-3706-4cca-88e9-3b524f6b840e)

A problem of the windows crate?

I recommend to specify the generic parameters of transmute. I recommend to read this lint description:

- https://rust-lang.github.io/rust-clippy/master/index.html#missing_transmute_annotations

Reply:
I don't think the `windows` crate is wrong _per se_ because `PWSTR` and `LPWSTR` are synonymous; they both resolve to a pointer to `WCHAR`.

What we see here is a quirk of Win32; where `LoadString*` has two modes of operation:

- Allocate a `WCHAR` of the size you want, pass the length in `cchBufferMax`; and if the resource is found, it will be copied into the buffer. Be careful of null termination and etc
- Pass "0" for `cchBufferMax` and the address of a `wchar_t*` to `lpBuffer`. and if the resource is found, you get back a read-only pointer to the string resource itself (not necessarily null-terminated, but we have the length in characters in the return value).

The second case obviously has a type mismatch. In C++ you would do something like this:

```
const wchar_t* pchStringBegin = nullptr;
int cchStringLength = ::LoadStringW(
        hInstance, 
        resourceId, 
        reinterpret_cast<PWSTR>(&pchStringBegin), 
        0);
```

I will look into the generic parameters for transmute, so thanks for that tip. But overall I'm hoping that the safety comment sufficiently explains the use of transmute here.

Reply:
Thank you for the explanation. Yes it’s good I think! The suspicious part is that it’s typically not a good idea to use `transmute` to forcibly change the type of something and there exist many better way of casting, but for this API very Rustophobic I don’t think we can do any better! Specifying the generic parameters will be the only thing I would ask for.